nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarConsumerProcessor.java

```diff
@@ -0,0 +1,296 @@
+    protected ExecutorService ackPool;
+    protected ExecutorCompletionService<Object> ackService;
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @OnScheduled
+    public void init(ProcessContext context) {
+       if (context.getProperty(ASYNC_ENABLED).isSet() && context.getProperty(ASYNC_ENABLED).asBoolean()) {
+           consumerPool = Executors.newFixedThreadPool(context.getProperty(MAX_ASYNC_REQUESTS).asInteger());
+           consumerService = new ExecutorCompletionService<>(consumerPool);
+           ackPool = Executors.newFixedThreadPool(context.getProperty(MAX_ASYNC_REQUESTS).asInteger() + 1);
+           ackService = new ExecutorCompletionService<>(ackPool);
+       }
+    }
+
+    @OnUnscheduled
+    public void shutDown(final ProcessContext context) {
+        if (context.getProperty(ASYNC_ENABLED).isSet() && context.getProperty(ASYNC_ENABLED).asBoolean()) {
+            try {
+                consumerPool.shutdown();
+                consumerPool.awaitTermination(10, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+                getLogger().error("Unable to stop all the Pulsar Consumers", e);
+            }
+
+            try {
+                ackPool.shutdown();
+                ackPool.awaitTermination(10, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+                getLogger().error("Unable to wait for all of the message acknowledgments to be sent", e);
+            }
+        }
+        close(context);
+    }
+
+    @OnStopped
+    public void close(final ProcessContext context) {
+        try {
+            getLogger().info("Disconnecting Pulsar Consumer");
+            if (consumer != null) {
+              consumer.close();
+            }
+        } catch (Exception e) {
+           getLogger().error("Unable to close Pulsar consumer", e);
+        } finally {
+           consumer = null;
+        }
+    }
+
+    protected void consumeAsync(ProcessContext context, ProcessSession session) throws PulsarClientException {
+        Consumer<T> consumer = getConsumer(context);
+
+        try {
+            consumerService.submit(() -> {
+               return consumer.receive();
+            });
+        } catch (final RejectedExecutionException ex) {
+            getLogger().error("Unable to consume aany more Pulsar messages", ex);
+        }
+    }
+
+    protected Consumer<T> getConsumer(ProcessContext context) throws PulsarClientException {
+
+        if (consumer == null) {
+           ConsumerBuilder<T> builder = (ConsumerBuilder<T>) getPulsarClient(context).newConsumer();
+
+           if (context.getProperty(TOPICS).isSet()) {
+              builder = builder.topic(context.getProperty(TOPICS)
+                               .evaluateAttributeExpressions().getValue().split("[, ]"));
```

Comment:
The call to split("[, ]") interprets its parameter as a regex, so the pattern '[, ]' means split on either a comma or a space... This allows the user to use a CSV list or a space-separated list. https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#split-java.lang.String-

Reply:
@david-streamlio yes, that is true. However, it does not allow for commas separating with spaces. I.e., 90% of the time users will enter "myTopic, yourTopic" and that regex will split it into "myTopic" followed by " yourTopic".... note the space before the 'y' in the second topic name. It is best to trim these out. Otherwise, users end up getting back error messages like "Could not find topic  yourTopic" and then it's not clear that there's a leading space and a lot of time is spent trying to debug.